backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -845,6 +853,59 @@ public FutureExperimentRunDAO(
         executor);
   }
 
+  private void checkVersionInputBasedOnPrivileges(
+      ExperimentRun.Builder experimentRunBuilder,
+      Set<Long> accessibleRepoIdsSet,
+      Set<Long> notAccessibleRepoIdIdsSet) {
+    Long repoId = experimentRunBuilder.getVersionedInputs().getRepositoryId();
+    if (accessibleRepoIdsSet.contains(repoId)) {
+      accessibleRepoIdsSet.add(repoId);
+    } else if (notAccessibleRepoIdIdsSet.contains(repoId)) {
+      notAccessibleRepoIdIdsSet.add(repoId);
+      experimentRunBuilder.clearVersionedInputs().build();
+    } else {
+      try {
+        Boolean isSelfAllowed =
+            IsRepositorySelfAllowed(
+                    Collections.singletonList(String.valueOf(repoId)),
+                    ModelDBActionEnum.ModelDBServiceActions.READ)
+                .get();
+        if (isSelfAllowed) {
+          accessibleRepoIdsSet.add(repoId);
+        } else {
+          experimentRunBuilder.clearVersionedInputs().build();
+          notAccessibleRepoIdIdsSet.add(repoId);
+        }
+      } catch (ExecutionException | InterruptedException ex) {
+        throw new ModelDBException(ex);
+      }
+    }
+  }
+
+  private InternalFuture<Boolean> IsRepositorySelfAllowed(
```

Comment:
We already have a method for this for any type. Let's re-use instead of coyping

Reply:
Removed and modify existing one for re-use